infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/Orchestrator.java

```diff
@@ -75,45 +91,47 @@ void handleQuery(CorfuPayloadMsg<OrchestratorRequest> msg, ChannelHandlerContext
                 .payloadMsg(new OrchestratorResponse(resp)));
     }
 
-    void addNode(CorfuPayloadMsg<OrchestratorRequest> msg, ChannelHandlerContext ctx, IServerRouter r) {
-        AddNodeRequest req = (AddNodeRequest) msg.getPayload().getRequest();
-        if (findUUID(req.getEndpoint()) != null) {
-            // An add node workflow is already executing for this endpoint, return
+    /**
+     *
+     * Dispatch a workflow create request.
+     *
+     * Create and start a workflow on this orchestrator, if there is
+     * an existing workflow that is executing on the same endpoint,
+     * then just return the corresponding workflow id. Dispatch is the only
+     * place where workflows are created based on reading activeWorkflows
+     * and therefore needs to be synchronized to prevent launching multiple
+     * workflows for the same endpoint concurrently.
+     *
+     * @param msg corfu message containing the create workflow request
+     * @param ctx netty ChannelHandlerContext
+     * @param r   server router
+     */
+    synchronized void dispatch(CorfuPayloadMsg<OrchestratorRequest> msg,
+                               ChannelHandlerContext ctx, IServerRouter r) {
+        CreateRequest req = (CreateRequest) msg.getPayload().getRequest();
+
+        UUID id = activeWorkflows.inverse().get(req.getEndpoint());
```

Comment:
This does not dedupe on workflow type as well. As you mentioned you will address this in a subsequent PR